### PR TITLE
[Eager] saved_tensors_hooks

### PR DESCRIPTION
**Pull request by wanghuancoder**
### PR types

New features
### PR changes

Others 
### Describe

添加paddle.autograd.saved_tensors_hooks模块

### CODE REVIEW

File: paddle/fluid/eager/CMakeLists.txt [hunk 1]
```diff
@@ -75,7 +76,17 @@ cc_library(
...
        autograd_meta
-       hook_utils)
+       hook_utils
+       tensor_wrapper)
+
+set(tensor_wrapper_deps phi_tensor)
+if(NOT ((NOT WITH_PYTHON) AND ON_INFER))
+  set(tensor_wrapper_deps ${tensor_wrapper_deps} python pybind)
+endif()
+cc_library(
```
Comment (JiabinYang) [inline paddle/fluid/eager/CMakeLists.txt:86]: Do we need this to be an independent cc_library or just make this compiled with other hooks?

File: paddle/fluid/eager/custom_operator/custom_operator_node.cc [hunk 2]
```diff
@@ -14,6 +14,7 @@
 
 #include "paddle/fluid/eager/custom_operator/custom_operator_node.h"
 
+#include "paddle/fluid/eager/utils.h"
```
Comment (JiabinYang) [inline paddle/fluid/eager/custom_operator/custom_operator_node.cc:17]: why this?

File: paddle/fluid/eager/tensor_wrapper.cc [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+#include "pybind11/pybind11.h"
+#include "pybind11/pytypes.h"
+#endif
+
+#if !(defined(PADDLE_NO_PYTHON) && defined(PADDLE_ON_INFERENCE))
+PyTypeObject* p_tensor_type_tensor_wrapper = nullptr;
+#endif
+namespace egr {
+
+#if !(defined(PADDLE_NO_PYTHON) && defined(PADDLE_ON_INFERENCE))
```
Comment (JiabinYang) [inline paddle/fluid/eager/tensor_wrapper.cc:43]: how about impl this into eager_utils.cc under pybind directory

File: paddle/fluid/eager/tensor_wrapper.cc [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+  weak_grad_node_ = other.weak_grad_node_;
+  inplace_version_snapshot_ = other.inplace_version_snapshot_;
+  packed_tensor_info_ = other.packed_tensor_info_;
+  unpack_hook_ = other.unpack_hook_;
+  Py_XINCREF(packed_tensor_info_);
+  Py_XINCREF(unpack_hook_);
+}
+
+TensorWrapper& TensorWrapper::operator=(const TensorWrapper& other) {
+  no_need_buffer_ = other.no_need_buffer_;
```
Comment (JiabinYang) [inline paddle/fluid/eager/tensor_wrapper.cc:80]: why not calling copy constructor before?

File: paddle/fluid/eager/tensor_wrapper.cc [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+          "Unrecognized tensor type for no_need_buffer feature"));
+    }
+  } else {
+#if !(defined(PADDLE_NO_PYTHON) && defined(PADDLE_ON_INFERENCE))
+    if (SavedTensorsHooks::GetInstance().is_enable() &&
+        tensor.is_dense_tensor()) {
+      phi::DenseTensor* dense_tensor =
+          static_cast<phi::DenseTensor*>(tensor.impl().get());
+      intermidiate_tensor_.set_impl(
+          std::move(std::make_shared<phi::DenseTensor>(
```
Comment (JiabinYang) [inline paddle/fluid/eager/tensor_wrapper.cc:138]: why this? It should not be an no need buffer hook.

File: paddle/fluid/eager/tensor_wrapper.cc [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+              std::make_shared<phi::Allocation>(nullptr, 0, tensor.place()),
+              dense_tensor->meta())));
+      auto pack_hook = SavedTensorsHooks::GetInstance().get_pack_hook();
```
Comment (JiabinYang) [inline paddle/fluid/eager/tensor_wrapper.cc]: why not make this into global Controller?

File: paddle/fluid/eager/saved_tensors_hooks.h [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace egr {
+
+class SavedTensorsHooks {
+ public:
+  SavedTensorsHooks() = default;
+
+  ~SavedTensorsHooks() {}
+
+  void set_hooks(PyObject* pack_hook, PyObject* unpack_hook) {
```
Comment (JiabinYang) [inline paddle/fluid/eager/saved_tensors_hooks.h]: Bad name, name this into register

File: paddle/fluid/eager/tensor_wrapper.cc [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+          std::move(std::make_shared<phi::DenseTensor>(
+              std::make_shared<phi::Allocation>(nullptr, 0, tensor.place()),
+              std::move(dense_tensor->meta()))));
+    } else {
+      PADDLE_THROW(paddle::platform::errors::Fatal(
+          "Unrecognized tensor type for no_need_buffer feature"));
+    }
+  } else {
+#if !(defined(PADDLE_NO_PYTHON) && defined(PADDLE_ON_INFERENCE))
+    if (SavedTensorsHooks::GetInstance().is_enable() &&
```
Comment (JiabinYang) [inline paddle/fluid/eager/tensor_wrapper.cc]: Seal this into a function called apply hook maybe

File: python/paddle/autograd/saved_tensors_hooks.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from paddle.fluid import core
+
+__all__ = []
+
+
+class saved_tensors_hooks():
```
Comment (JiabinYang) [inline python/paddle/autograd/saved_tensors_hooks.py]: class named in Upper case.

**Review by JiabinYang — COMMENTED** [review]
some comments

File: paddle/fluid/eager/custom_operator/custom_operator_node.cc (hunk 2 above)
Comment (wanghuancoder) [reply paddle/fluid/eager/custom_operator/custom_operator_node.cc:17]: 当tensor_wrapper.h中的include挪到tensor_wrapper.cc中以后，这个文件缺少utils.h

File: paddle/fluid/eager/CMakeLists.txt (hunk 1 above)
Comment (wanghuancoder) [reply paddle/fluid/eager/CMakeLists.txt:86]: 独立编译cc_library是因为，这样不需要所有include这个tensor_wrapper.h得cc都添加python pybind

File: paddle/fluid/eager/tensor_wrapper.cc (hunk 3 above)
Comment (wanghuancoder) [reply paddle/fluid/eager/tensor_wrapper.cc:43]: tensor_wrapper.cc是底层模块，而eager_utils.cc是辅助于pybind下，向python暴露接口使用的。
在编译时，是先编译tensor_wrapper.cc，最后在打包paddle_pybind.so时才会编译eager_utils.cc。
受限于咱们cmake的约束。如果改变编译先后顺序会面临很多问题。我试过。放弃了。

File: paddle/fluid/eager/tensor_wrapper.cc (hunk 4 above)
Comment (wanghuancoder) [reply paddle/fluid/eager/tensor_wrapper.cc:80]: why not calling copy constructor before? 这是啥意思？为啥不调用构造函数？
通常我们反向node中的逻辑都是这么写的。先构造，再赋值。赋值涉及copy。
```
  void SetTensorWrapperx(const paddle::experimental::Tensor& x) {
    x_ = egr::TensorWrapper(x, false);
  }
```

File: paddle/fluid/eager/tensor_wrapper.cc (hunk 5 above)
Comment (wanghuancoder) [reply paddle/fluid/eager/tensor_wrapper.cc:138]: 这里是把dense_tensor的meta信息先保存下来。只让用户负责save data就行了，恢复也只恢复data

File: paddle/fluid/eager/tensor_wrapper.cc (hunk 6 above)
Comment (wanghuancoder) [reply paddle/fluid/eager/tensor_wrapper.cc]: 啥意思？global Controller是啥？

**Pull request closed without merging**